File: src/lib/wow/db/dbd-parser.ts
/**
 * DBD (database definition) parser, ported from wow.export (src/js/db/DBDParser.js).
 */
/* eslint-disable max-classes-per-file */
import type { BufferWrapper } from '../formats/buffer';

/** Pattern to match column definitions in a DBD document. */
const PATTERN_COLUMN = /^(int|float|locstring|string)(<[^:]+::[^>]+>)?\s([^\s]+)/;

/** Pattern to match build identifiers in a DBD document. */
const PATTERN_BUILD = /^BUILD\s(.*)/;

/** Pattern to match build range identifiers in a DBD document. */
const PATTERN_BUILD_RANGE = /([^-]+)-(.*)/;

/** Pattern to match comment entries in a DBD document. */
const PATTERN_COMMENT = /^COMMENT\s/;

/** Pattern to match layout hash identifiers in a DBD document. */
const PATTERN_LAYOUT = /^LAYOUT\s(.*)/;

/** Pattern to match a field entry in a DBD document. */
const PATTERN_FIELD = /^(\$([^$]+)\$)?([^<[]+)(<(u|)(\d+)>)?(\[(\d+)\])?$/;

/** Pattern to match the components of a build ID. */
const PATTERN_BUILD_ID = /(\d+).(\d+).(\d+).(\d+)/;

export type DBDColumnType = 'int' | 'float' | 'locstring' | 'string';

interface BuildID {
  major: number;
  minor: number;
  patch: number;
  rev: number;
}

/** Parse a build ID into components. */
function parseBuildID(buildID: string): BuildID {
  const parts = buildID.match(PATTERN_BUILD_ID);
  const entry: BuildID = {
    major: 0, minor: 0, patch: 0, rev: 0,
  };

  if (parts !== null) {
    entry.major = parseInt(parts[1], 10);
    entry.minor = parseInt(parts[2], 10);
    entry.patch = parseInt(parts[3], 10);
    entry.rev = parseInt(parts[4], 10);
  }

  return entry;
}

/** Returns true if the provided build falls within the provided range. */
function isBuildInRange(buildStr: string, minStr: string, maxStr: string): boolean {
  const build = parseBuildID(buildStr);
  const min = parseBuildID(minStr);
  const max = parseBuildID(maxStr);

  if (build.major < min.major || build.major > max.major) return false;
  if (build.minor < min.minor || build.minor > max.minor) return false;
  if (build.patch < min.patch || build.patch > max.patch) return false;
  if (build.rev < min.rev || build.rev > max.rev) return false;

  return true;
}

export class DBDField {
  type: DBDColumnType;

  name: string;

  isSigned = true;

  isID = false;

  isInline = true;

  isRelation = false;

  arrayLength = -1;

  size = -1;

  constructor(fieldName: string, fieldType: DBDColumnType) {
    this.type = fieldType;
    this.name = fieldName;
  }
}

export class DBDEntry {
  builds = new Set<string>();

  buildRanges = new Set<{ min: string; max: string }>();

  layoutHashes = new Set<string>();

  fields = new Set<DBDField>();

  /** Add a build to this DBD entry. */
  addBuild(min: string, max?: string): void {
    if (max !== undefined) this.buildRanges.add({ min, max });
    else this.builds.add(min);
  }

  /** Add a layout hash to this DBD entry. */
  addLayoutHashes(...hashes: string[]): void {
    for (const hash of hashes) this.layoutHashes.add(hash);
  }

  /** Add a field to this DBD entry. */
  addField(field: DBDField): void {
    this.fields.add(field);
  }

  /** Check if this entry is valid for the provided buildID or layout hash. */
  isValidFor(buildID: string, layoutHash: string): boolean {
    // Layout hash takes priority, being the quickest to check.
    if (this.layoutHashes.has(layoutHash)) return true;

    // Check for a single build ID.
    if (this.builds.has(buildID)) return true;

    // Fallback to checking build ranges.
    for (const range of this.buildRanges) {
      if (isBuildInRange(buildID, range.min, range.max)) return true;
    }

    return false;
  }
}

export class DBDParser {
  entries = new Set<DBDEntry>();

  columns = new Map<string, DBDColumnType>();

  constructor(data: BufferWrapper) {
    this.parse(data);
  }

  /** Get a DBD structure for the provided buildID and layoutHash. */
  getStructure(buildID: string, layoutHash: string): DBDEntry | null {
    for (const entry of this.entries) {
      if (entry.isValidFor(buildID, layoutHash)) return entry;
    }

    return null;
  }

  /** Parse the contents of a DBD document. */
  private parse(data: BufferWrapper): void {
    const lines = data.readLines();

    // Separate the file into chunks separated by empty lines.
    let chunk: string[] = [];
    for (const line of lines) {
      if (line.trim().length > 0) {
        chunk.push(line);
      } else {
        this.parseChunk(chunk);
        chunk = [];
      }
    }

    // Ensure last chunk is accounted for.
    if (chunk.length > 0) this.parseChunk(chunk);

    if (this.columns.size === 0) throw new Error('Invalid DBD: No columns defined.');
  }

  /** Parse a chunk from this DBD document. */
  private parseChunk(chunk: string[]): void {
    if (chunk[0] === 'COLUMNS') {
      this.parseColumnChunk(chunk);
    } else {
      const entry = new DBDEntry();
      for (const line of chunk) {
        // Build IDs.
        const buildMatch = line.match(PATTERN_BUILD);
        if (buildMatch !== null) {
          // BUILD 1.7.0.4671-1.8.0.4714
          // BUILD 0.9.1.3810
          // BUILD 1.13.6.36231, 1.13.6.36310
          const builds = buildMatch[1].split(',');
          for (const build of builds) {
            const buildRange = build.match(PATTERN_BUILD_RANGE);
            if (buildRange !== null) entry.addBuild(buildRange[1], buildRange[2]);
            else entry.addBuild(build.trim());
          }

          continue;
        }

        // Skip comments.
        if (line.match(PATTERN_COMMENT) !== null) continue;

        // Layout hashes.
        const layoutMatch = line.match(PATTERN_LAYOUT);
        if (layoutMatch !== null) {
          // LAYOUT 0E84A21C, 35353535
          entry.addLayoutHashes(...(layoutMatch[1].split(',').map((e) => e.trim())));
          continue;
        }

        const fieldMatch = line.match(PATTERN_FIELD);
        if (fieldMatch !== null) {
          const fieldName = fieldMatch[3];
          const fieldType = this.columns.get(fieldName);

          if (fieldType === undefined) throw new Error(`Invalid DBD: No field type defined for ${fieldName}`);

          const field = new DBDField(fieldName, fieldType);

          // Parse annotations, (eg 'id,noninline,relation').
          if (fieldMatch[2] !== undefined) {
            const annotations = fieldMatch[2].split(',');
            if (annotations.includes('id')) field.isID = true;
            if (annotations.includes('noninline')) field.isInline = false;
            if (annotations.includes('relation')) field.isRelation = true;
          }

          // Parse signedness, either 'u' or undefined.
          if (fieldMatch[5] && fieldMatch[5].length > 0) field.isSigned = false;

          // Parse data size (eg '32').
          if (fieldMatch[6] !== undefined) {
            const dataSize = parseInt(fieldMatch[6], 10);
            if (!Number.isNaN(dataSize)) field.size = dataSize;
          }

          // Parse array size (eg '2').
          if (fieldMatch[8] !== undefined) {
            const arrayLength = parseInt(fieldMatch[8], 10);
            if (!Number.isNaN(arrayLength)) field.arrayLength = arrayLength;
          }

          entry.addField(field);
        }
      }

      this.entries.add(entry);
    }
  }

  /** Parse the column definition of a DBD document. */
  private parseColumnChunk(chunk: string[]): void {
    if (chunk === undefined) throw new Error('Invalid DBD: Missing column definitions.');

    // Remove the COLUMNS header.
    chunk.shift();

    for (const entry of chunk) {
      const match = entry.match(PATTERN_COLUMN);
      if (match !== null) {
        const columnType = match[1] as DBDColumnType; // int|float|locstring|string
        // const columnForeignKey = match[2]; // <TableName::ColumnName> or undefined
        const columnName = match[3].replace('?', ''); // Field_6_0_1_18179_000?

        this.columns.set(columnName, columnType);
      }
    }
  }
}

export default DBDParser;

File: golang/internal/wow/db/dbd_parser.go
package db

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/pqhuy98/wow-converter/internal/buffer"
)

var (
	patternColumn    = regexp.MustCompile(`^(int|float|locstring|string)(<[^:]+::[^>]+>)?\s([^\s]+)`)
	patternBuild     = regexp.MustCompile(`^BUILD\s(.*)`)
	patternBuildRange = regexp.MustCompile(`([^-]+)-(.*)`)
	patternComment   = regexp.MustCompile(`^COMMENT\s`)
	patternLayout    = regexp.MustCompile(`^LAYOUT\s(.*)`)
	patternField     = regexp.MustCompile(`^(\$([^$]+)\$)?([^<[]+)(<(u|)(\d+)>)?(\[(\d+)\])?$`)
	patternBuildID   = regexp.MustCompile(`(\d+).(\d+).(\d+).(\d+)`)
)

// DBDColumnType is a column type from a DBD COLUMNS section.
type DBDColumnType string

const (
	DBDColumnInt       DBDColumnType = "int"
	DBDColumnFloat     DBDColumnType = "float"
	DBDColumnLocString DBDColumnType = "locstring"
	DBDColumnString    DBDColumnType = "string"
)

type buildID struct {
	major, minor, patch, rev int
}

func parseBuildID(build string) buildID {
	parts := patternBuildID.FindStringSubmatch(build)
	entry := buildID{}
	if parts != nil {
		entry.major, _ = strconv.Atoi(parts[1])
		entry.minor, _ = strconv.Atoi(parts[2])
		entry.patch, _ = strconv.Atoi(parts[3])
		entry.rev, _ = strconv.Atoi(parts[4])
	}
	return entry
}

func isBuildInRange(buildStr, minStr, maxStr string) bool {
	build := parseBuildID(buildStr)
	min := parseBuildID(minStr)
	max := parseBuildID(maxStr)
	if build.major < min.major || build.major > max.major {
		return false
	}
	if build.minor < min.minor || build.minor > max.minor {
		return false
	}
	if build.patch < min.patch || build.patch > max.patch {
		return false
	}
	if build.rev < min.rev || build.rev > max.rev {
		return false
	}
	return true
}

// DBDField describes a field in a DBD entry.
type DBDField struct {
	Type        DBDColumnType
	Name        string
	IsSigned    bool
	IsID        bool
	IsInline    bool
	IsRelation  bool
	ArrayLength int
	Size        int
}

// DBDEntry is a version-specific table definition from a DBD file.
type DBDEntry struct {
	Builds      map[string]struct{}
	BuildRanges []struct{ Min, Max string }
	LayoutHashes map[string]struct{}
	Fields      []*DBDField
}

func newDBDEntry() *DBDEntry {
	return &DBDEntry{
		Builds:       make(map[string]struct{}),
		LayoutHashes: make(map[string]struct{}),
	}
}

func (e *DBDEntry) addBuild(min string, max ...string) {
	if len(max) > 0 {
		e.BuildRanges = append(e.BuildRanges, struct{ Min, Max string }{min, max[0]})
	} else {
		e.Builds[min] = struct{}{}
	}
}

func (e *DBDEntry) addLayoutHashes(hashes ...string) {
	for _, h := range hashes {
		e.LayoutHashes[strings.TrimSpace(h)] = struct{}{}
	}
}

func (e *DBDEntry) addField(field *DBDField) {
	e.Fields = append(e.Fields, field)
}

// IsValidFor reports whether this entry matches buildID or layoutHash.
func (e *DBDEntry) IsValidFor(buildID, layoutHash string) bool {
	if _, ok := e.LayoutHashes[layoutHash]; ok {
		return true
	}
	if _, ok := e.Builds[buildID]; ok {
		return true
	}
	for _, r := range e.BuildRanges {
		if isBuildInRange(buildID, r.Min, r.Max) {
			return true
		}
	}
	return false
}

// DBDParser parses DBD definition files.
type DBDParser struct {
	Entries []*DBDEntry
	columns map[string]DBDColumnType
}

// NewDBDParser parses DBD data from a buffer.
func NewDBDParser(data *buffer.Buffer) (*DBDParser, error) {
	p := &DBDParser{columns: make(map[string]DBDColumnType)}
	if err := p.parse(data); err != nil {
		return nil, err
	}
	return p, nil
}

// GetStructure returns the matching DBD entry for build/layout.
func (p *DBDParser) GetStructure(buildID, layoutHash string) *DBDEntry {
	for _, entry := range p.Entries {
		if entry.IsValidFor(buildID, layoutHash) {
			return entry
		}
	}
	return nil
}

func (p *DBDParser) parse(data *buffer.Buffer) error {
	lines := data.ReadLines("utf8")
	var chunk []string
	for _, line := range lines {
		if strings.TrimSpace(line) != "" {
			chunk = append(chunk, line)
		} else {
			if err := p.parseChunk(chunk); err != nil {
				return err
			}
			chunk = nil
		}
	}
	if len(chunk) > 0 {
		if err := p.parseChunk(chunk); err != nil {
			return err
		}
	}
	if len(p.columns) == 0 {
		return fmt.Errorf("invalid DBD: no columns defined")
	}
	return nil
}

func (p *DBDParser) parseChunk(chunk []string) error {
	if len(chunk) == 0 {
		return nil
	}
	if chunk[0] == "COLUMNS" {
		return p.parseColumnChunk(chunk)
	}
	entry := newDBDEntry()
	for _, line := range chunk {
		if buildMatch := patternBuild.FindStringSubmatch(line); buildMatch != nil {
			for _, build := range strings.Split(buildMatch[1], ",") {
				build = strings.TrimSpace(build)
				if buildRange := patternBuildRange.FindStringSubmatch(build); buildRange != nil {
					entry.addBuild(buildRange[1], buildRange[2])
				} else {
					entry.addBuild(build)
				}
			}
			continue
		}
		if patternComment.MatchString(line) {
			continue
		}
		if layoutMatch := patternLayout.FindStringSubmatch(line); layoutMatch != nil {
			entry.addLayoutHashes(strings.Split(layoutMatch[1], ",")...)
			continue
		}
		if fieldMatch := patternField.FindStringSubmatch(line); fieldMatch != nil {
			fieldName := fieldMatch[3]
			fieldType, ok := p.columns[fieldName]
			if !ok {
				return fmt.Errorf("invalid DBD: no field type defined for %s", fieldName)
			}
			field := &DBDField{
				Type:        fieldType,
				Name:        fieldName,
				IsSigned:    true,
				IsInline:    true,
				ArrayLength: -1,
				Size:        -1,
			}
			if fieldMatch[2] != "" {
				for _, ann := range strings.Split(fieldMatch[2], ",") {
					switch ann {
					case "id":
						field.IsID = true
					case "noninline":
						field.IsInline = false
					case "relation":
						field.IsRelation = true
					}
				}
			}
			if fieldMatch[5] != "" {
				field.IsSigned = false
			}
			if fieldMatch[6] != "" {
				if size, err := strconv.Atoi(fieldMatch[6]); err == nil {
					field.Size = size
				}
			}
			if fieldMatch[8] != "" {
				if n, err := strconv.Atoi(fieldMatch[8]); err == nil {
					field.ArrayLength = n
				}
			}
			entry.addField(field)
		}
	}
	p.Entries = append(p.Entries, entry)
	return nil
}

func (p *DBDParser) parseColumnChunk(chunk []string) error {
	if len(chunk) == 0 {
		return fmt.Errorf("invalid DBD: missing column definitions")
	}
	for _, entry := range chunk[1:] {
		if match := patternColumn.FindStringSubmatch(entry); match != nil {
			columnName := strings.ReplaceAll(match[3], "?", "")
			p.columns[columnName] = DBDColumnType(match[1])
		}
	}
	return nil
}
